File: python_files/steering_aligner2.py
#!/usr/bin/env python3
"""
Steering Aligner Node — ROS2 Humble

📘 Description:
Smoothly aligns the steering to a target heading using Exponential Moving Average (EMA).
After alignment (within tolerance), it holds and stabilizes the heading for a short duration
before sending a Boolean flag on `/alignment_done` = True. During the hold phase,
small corrections are made to ensure perfect heading stability before handing control
to the Stanley controller.

💡 Topics:
  - Subscribes:
      /aligned (std_msgs/Float32) → Target heading [degrees]
      /jmoab_compass (std_msgs/Float32MultiArray) → [roll, pitch, heading]
  - Publishes:
      /steering_angle (std_msgs/Float32)
      /alignment_done (std_msgs/Bool)
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32, Float32MultiArray, Bool
import math
import time
from rclpy.qos import QoSProfile, ReliabilityPolicy, QoSDurabilityPolicy


class SteeringAligner(Node):
    def __init__(self):
        super().__init__('steering_aligner')

        # --------------------------------------------------
        # Parameters
        # --------------------------------------------------
        self.tolerance_deg = 2.0             # ± tolerance in degrees
        self.max_pwm = 17
        self.min_pwm = 0.0
        self.ema_alpha = 0.5                 # smoothing factor (0.0–1.0)
        self.hold_duration = 0.26             # seconds to hold and stabilize after alignment
        self.disable_time = 0.0              # seconds to pause after alignment
        self.last_disable_time = 0.0

        # --------------------------------------------------
        # Internal state
        # --------------------------------------------------
        self.target_heading = None
        self.current_heading = None
        self.smoothed_pwm = 0.0
        self.active = True                   # Whether aligner is actively controlling
        self.holding = False                 # Whether in holding/stabilization phase
        self.hold_start_time = None

        # --------------------------------------------------
        # Publishers
        # --------------------------------------------------
        qos_policy = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=QoSDurabilityPolicy.VOLATILE,
            depth=1
        )
        self.steer_pub = self.create_publisher(Float32, '/steering_angle', qos_policy)
        self.align_done_pub = self.create_publisher(Bool, '/alignment_done', 10)

        # --------------------------------------------------
        # Subscribers
        # --------------------------------------------------
        self.create_subscription(Float32, '/aligned', self.aligned_callback, 10)
        self.create_subscription(Float32MultiArray, '/jmoab_compass', self.compass_callback, 10)

        # --------------------------------------------------
        # Timer
        # --------------------------------------------------
        self.create_timer(0.005, self.control_loop)
        self.get_logger().info("🧭 Steering Aligner Node started.")

    # ======================================================
    #  Callbacks
    # ======================================================
    def aligned_callback(self, msg: Float32):
        """New target heading (degrees)"""
        if time.time() - self.last_disable_time < self.disable_time:
            return  # Ignore if still in cooldown
        self.target_heading = msg.data % 360.0
        self.active = True
        self.holding = False
        self.get_logger().info(f"🎯 New target heading received: {self.target_heading:.2f}°")

    def compass_callback(self, msg: Float32MultiArray):
        """Current heading from compass (degrees)"""
        if len(msg.data) >= 3:
            self.current_heading = msg.data[2] % 360.0

    # ======================================================
    #  Helper Functions
    # ======================================================
    def angle_diff(self, a, b):
        """Smallest signed difference between two headings (degrees)"""
        d = (a - b + 180.0) % 360.0 - 180.0
        return d

    def apply_ema(self, new_val):
        """Smooths PWM updates"""
        self.smoothed_pwm = (self.ema_alpha * new_val) + (1 - self.ema_alpha) * self.smoothed_pwm
        return self.smoothed_pwm

    # ===========================================================
    #  Control Loop
    # ===========================================================
    def control_loop(self):
        if self.current_heading is None or self.target_heading is None:
            return

        # Skip if not active or not holding
        if not self.active and not self.holding:
            return

        # Compute heading error
        error = self.angle_diff(self.target_heading, self.current_heading)

        # ==================================================
        # 1️⃣ ALIGNING PHASE
        # ==================================================
        if not self.holding:
            if abs(error) <= self.tolerance_deg:
                # Enter holding phase
                self.holding = True
                self.hold_start_time = time.time()
                self.get_logger().info(
                    f"✅ Aligned! Holding for {self.hold_duration:.1f}s to stabilize..."
                )
                return
            else:
                # Keep aligning
                pwm_cmd = (abs(error) / 2.0) * self.max_pwm
                pwm_cmd = min(max(pwm_cmd, self.min_pwm), self.max_pwm)
                if error < 0:
                    pwm_cmd = -pwm_cmd
                pwm_cmd = self.apply_ema(pwm_cmd)
                self.publish_pwm(pwm_cmd)

        # ==================================================
        # 2️⃣ HOLDING PHASE (STABILIZATION)
        # ==================================================
        elif self.holding:
            hold_elapsed = time.time() - self.hold_start_time

            # If still within hold time
            if hold_elapsed < self.hold_duration:
                # Keep small corrections
                pwm_cmd = (abs(error) / 2.0) * self.max_pwm
                pwm_cmd = min(max(pwm_cmd, self.min_pwm), self.max_pwm)
                if error < 0:
                    pwm_cmd = -pwm_cmd
                pwm_cmd = self.apply_ema(pwm_cmd)
                self.publish_pwm(pwm_cmd)
                return

            # ==================================================
            # 3️⃣ FINALIZE AFTER HOLD
            # ==================================================
            else:
                self.publish_alignment_done()
                self.active = False
                self.holding = False
                self.last_disable_time = time.time()
                self.get_logger().info(
                    f"📡 Alignment complete and stabilized — released control."
                )

    # ======================================================
    #  Publishers
    # ======================================================
    def publish_pwm(self, pwm_value):
        msg = Float32()
        msg.data = pwm_value
        self.steer_pub.publish(msg)

    def publish_alignment_done(self):
        msg = Bool()
        msg.data = True
        self.align_done_pub.publish(msg)
        self.get_logger().info("📡 Published /alignment_done = True")

# ==========================================================
#  Main
# ==========================================================
def main(args=None):
    rclpy.init(args=args)
    node = SteeringAligner()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("🛑 Steering Aligner interrupted by user.")
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/steering_aligner2.cpp
// ==========================================
// Basic Libraries & Headers
// ==========================================
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/qos.hpp>
#include <std_msgs/msg/float32.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>
#include <std_msgs/msg/bool.hpp>
#include "heading_controller_pkg/steering_aligner2_utils.hpp"

// ==========================================
// Type Aliases for Standard ROS 2 Messages
// ==========================================
using Float32 = std_msgs::msg::Float32;
using Float32MultiArray = std_msgs::msg::Float32MultiArray;
using Bool = std_msgs::msg::Bool;

// ==========================================
// Steering Aligner Node Class (simpler variant)
// Proportional alignment with EMA smoothing, then a short hold/stabilize
// phase before publishing /alignment_done. No adaptive gain, slew
// limiting, or oscillation detection -- see dynamic_steering_aligner.cpp
// for the more advanced version.
//
// NOTE: this node uses the SAME ROS node name ("steering_aligner") as
// dynamic_steering_aligner.cpp. Running both at once will fail with a
// duplicate-node-name error -- remap one at launch if you need both,
// e.g. --ros-args -r __node:=steering_aligner2.
//
// (angle_diff() / apply_ema() / compute_aligner2_step() now live in
// steering_aligner2_utils.hpp so they can be unit-tested directly with
// gtest, without needing a running rclcpp::Node)
// ==========================================
class SteeringAligner : public rclcpp::Node {
public:
    SteeringAligner() : Node("steering_aligner") {
        // ----------------------------------------------------
        // Declare ROS 2 Parameters (defaults match the Python node)
        // ----------------------------------------------------
        this->declare_parameter<float>("tolerance_deg", 2.0f);
        this->declare_parameter<float>("max_pwm", 17.0f);
        this->declare_parameter<float>("min_pwm", 0.0f);
        this->declare_parameter<float>("ema_alpha", 0.5f);
        this->declare_parameter<float>("hold_duration", 0.26f);
        this->declare_parameter<float>("disable_time", 0.0f);

        // ----------------------------------------------------
        // Load Parameters into Member Variables
        // ----------------------------------------------------
        params_.tolerance_deg = static_cast<float>(this->get_parameter("tolerance_deg").as_double());
        params_.max_pwm = static_cast<float>(this->get_parameter("max_pwm").as_double());
        params_.min_pwm = static_cast<float>(this->get_parameter("min_pwm").as_double());
        params_.ema_alpha = static_cast<float>(this->get_parameter("ema_alpha").as_double());
        params_.hold_duration = static_cast<float>(this->get_parameter("hold_duration").as_double());
        params_.disable_time = static_cast<double>(this->get_parameter("disable_time").as_double());

        // ----------------------------------------------------
        // Setup Publishers
        // ----------------------------------------------------
        rclcpp::QoS qos_policy(1);
        qos_policy.reliability(rclcpp::ReliabilityPolicy::Reliable);
        qos_policy.durability(rclcpp::DurabilityPolicy::Volatile);

        steer_pub_ = this->create_publisher<Float32>("/steering_angle", qos_policy);
        // Matches the Python: align_done_pub uses default QoS (depth 10),
        // NOT the custom qos_policy used for steer_pub.
        align_done_pub_ = this->create_publisher<Bool>("/alignment_done", 10);

        // ----------------------------------------------------
        // Setup Subscriptions
        // ----------------------------------------------------
        aligned_sub_ = this->create_subscription<Float32>(
            "/aligned", 10, std::bind(&SteeringAligner::aligned_callback, this, std::placeholders::_1));
        compass_sub_ = this->create_subscription<Float32MultiArray>(
            "/jmoab_compass", 10, std::bind(&SteeringAligner::compass_callback, this, std::placeholders::_1));

        // ----------------------------------------------------
        // Initialize State
        // ----------------------------------------------------
        state_ = Aligner2State{};

        // ----------------------------------------------------
        // Setup Control Loop Timer (200Hz, matching the Python's 0.005s period)
        // ----------------------------------------------------
        timer_ = this->create_wall_timer(
            std::chrono::duration<double>(0.005),
            std::bind(&SteeringAligner::control_loop, this));

        RCLCPP_INFO(this->get_logger(), "🧭 Steering Aligner Node started.");
    }

private:
    // ==========================================
    // Callbacks
    // ==========================================
    void aligned_callback(const Float32::SharedPtr msg) {
        double now = this->now().seconds();
        if (now - state_.last_disable_time < params_.disable_time) {
            return;  // Ignore if still in cooldown
        }
        state_.target_heading = wrap360(msg->data);
        state_.active = true;
        state_.holding = false;
        RCLCPP_INFO(this->get_logger(), "🎯 New target heading received: %.2f°", state_.target_heading.value());
    }

    void compass_callback(const Float32MultiArray::SharedPtr msg) {
        if (msg->data.size() >= 3) {
            state_.current_heading = wrap360(msg->data[2]);
        }
    }

    // ==========================================
    // Control Loop
    // ==========================================
    void control_loop() {
        double now = this->now().seconds();
        Aligner2StepResult result = compute_aligner2_step(state_, now, params_);

        if (result.entered_hold) {
            RCLCPP_INFO(this->get_logger(), "✅ Aligned! Holding for %.1fs to stabilize...", params_.hold_duration);
            return;
        }

        if (result.should_publish_pwm) {
            publish_pwm(result.pwm_value);
            return;
        }

        if (result.alignment_done) {
            publish_alignment_done();
            RCLCPP_INFO(this->get_logger(), "📡 Alignment complete and stabilized — released control.");
            return;
        }
    }

    // ==========================================
    // Publishers
    // ==========================================
    void publish_pwm(float pwm_value) {
        Float32 msg;
        msg.data = pwm_value;
        steer_pub_->publish(msg);
    }

    void publish_alignment_done() {
        Bool msg;
        msg.data = true;
        align_done_pub_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "📡 Published /alignment_done = True");
    }

    // ==========================================
    // ROS 2 Communication Handles
    // ==========================================
    rclcpp::Publisher<Float32>::SharedPtr steer_pub_;
    rclcpp::Publisher<Bool>::SharedPtr align_done_pub_;
    rclcpp::Subscription<Float32>::SharedPtr aligned_sub_;
    rclcpp::Subscription<Float32MultiArray>::SharedPtr compass_sub_;
    rclcpp::TimerBase::SharedPtr timer_;

    // ==========================================
    // Configuration Parameters
    // ==========================================
    Aligner2Params params_;

    // ==========================================
    // Internal State
    // ==========================================
    Aligner2State state_;
};

// ==========================================
// Main Execution Entry Point
// ==========================================
int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SteeringAligner>();

    try {
        rclcpp::spin(node);
    } catch (const std::exception& e) {
        RCLCPP_INFO(node->get_logger(), "🛑 Steering Aligner interrupted by user.");
    }

    rclcpp::shutdown();
    return 0;
}